### PR TITLE
CR-1100458: multi device fix poll

### PR DESCRIPTION
**Pull request by sarab96**
Fix for multiple device to check cu completion on multiple devices

### CODE REVIEW

**Review by maxzhen — APPROVED** [review]

**Review by maxzhen — APPROVED** [review]

File: src/xma/src/xmaapi/xmaapi.cpp [hunk 1]
```diff
@@ -258,24 +260,32 @@ void xma_thread2() {
...
+
+    while (!g_xma_singleton->xma_exit) {
         if (g_xma_singleton->cpu_mode == XMA_CPU_MODE2) {
             std::this_thread::sleep_for(std::chrono::milliseconds(3));
+        } else if (g_xma_singleton->cpu_mode == XMA_CPU_MODE1) {
+            ::poll(all_pollfd.data(), all_pollfd.size(), 500);
+            //xclExecWait(priv2->dev_handle, 100);
         } else {
-            xclExecWait(priv2->dev_handle, 100);
+            std::this_thread::sleep_for(std::chrono::milliseconds(100));
```
Comment (stsoe) [inline src/xma/src/xmaapi/xmaapi.cpp:287]: What is this third case?   It  no longer calls xclExecWait, what happens in cpu_mode is not MODE2 or MODE1?

File: src/xma/src/xmaapi/xmaapi.cpp [hunk 2]
```diff
@@ -258,24 +260,32 @@ void xma_thread2() {
...
+        if (!drv) {
+            xma_logmsg(XMA_ERROR_LOG, XMAAPI_MOD, "Device handle error: %d", hw_device.dev_index);
+            return;
         }
-        XmaHwSessionPrivate *priv2 = (XmaHwSessionPrivate*) g_xma_singleton->all_sessions_vec[session_index].hw_session.private_do_not_use;
+        all_pollfd.emplace_back(pollfd{});
+        auto& tmp_fd = all_pollfd.back();
+        tmp_fd.fd = (int) drv->xclGetFd();
+        tmp_fd.events = POLLIN;
+        tmp_fd.revents = 0;
```
Comment (stsoe) [inline src/xma/src/xmaapi/xmaapi.cpp:277]: This is a bit backwards.    Typically one would do:
`all_polldf.emplace_back(pollfd{drv->xclGetFd(), POLLIN, 0});`

File: src/xma/src/xmaapi/xmaapi.cpp [hunk 3]
```diff
@@ -258,24 +260,32 @@ void xma_thread2() {
...
+    }
+
+    while (!g_xma_singleton->xma_exit) {
         if (g_xma_singleton->cpu_mode == XMA_CPU_MODE2) {
             std::this_thread::sleep_for(std::chrono::milliseconds(3));
+        } else if (g_xma_singleton->cpu_mode == XMA_CPU_MODE1) {
+            ::poll(all_pollfd.data(), all_pollfd.size(), 500);
```
Comment (stsoe) [inline src/xma/src/xmaapi/xmaapi.cpp:284]: Just one question.  Are the collected device fds guaranteed to all be valid while this thread is running?

File: src/xma/src/xmaapi/xmaapi.cpp [hunk 4]
```diff
@@ -258,24 +260,32 @@ void xma_thread2() {
...
-            std::this_thread::sleep_for(std::chrono::milliseconds(30));
-            continue;
-        }
-        if (session_index >= num_sessions) {
-            session_index = 0;
+
+    std::vector<pollfd> all_pollfd;
+    all_pollfd.reserve(MAX_XILINX_DEVICES);
+
+    for (XmaHwDevice& hw_device: g_xma_singleton->hwcfg.devices) {
```
Comment (stsoe) [inline src/xma/src/xmaapi/xmaapi.cpp:267]: Is g_xma_singleton->hwcfg.devices  unchanged for the lifetime of this thread?

**Review by stsoe — CHANGES_REQUESTED** [review]

**Comment by maxzhen** [comment]
Not going to merge.

**Pull request closed without merging**